### ISSUE TITLE
Build system and online download builder

### ISSUE THREAD

**Issue by arschmitz** [jquery/css-chassis#17]
We will want to make sure we have both an easy to use command line build system as well as an online download builder to allow people to pick just the modules they want. The online builder is important since we want to be accessible to designers who will shy away from a command prompt.

**Comment by mendeza** [comment]
I agree with Dave as to _why not both_, though I personally prefer maximum flexibility on the commandline. Downloading a custom build (and then hacking on it more) incurs a maintenance nightmare, arguably a form of technical debt. Despite extra weight, here at Thomson Reuters we often use an override approach to keep OSS source **completely** untouched, for the express purpose of easing updates. We have separate "theme" preprocessing that imports variables and mixins from the third party code so that we can "play with the same deck," so to speak, while keeping the third party in a pristine state.

**Comment by trailofdad** [comment]
I like the idea of having both CLI and a web builder. It caters to more than just one group of developers AND you get to make your build custom and lightweight (only the chassis). @rxaviers there are so many users of frameworks out there who don't care to use a command line and hate having bloated file systems (those tend to be the more design-centric developmers). Why not make using a framework more accessible to those who are casual developers? It in no way should affect the power users out there.

**Comment by rxaviers** [comment]
@trailofdad, making it more accessible is the idea. The problem is how. I'm not defending CLI is all we should have. I'm concerned that we provide + encourage a broken solution. A download builder is a bad habit IMO for maintenance purposes as discussed above.

**Comment by rxaviers** [comment]
> I imagine a system, similar to jQuery's, where urls with query params could be used to hit an api that would return the customized CSS. This would have the ability to cache output for subsequent api calls. The web based tool would just configure the api call.

@kristoferjoseph, this is an interesting approach. It's like having css-chassis as a service. Although, technically it's very similar to a Download Builder plus being a CDN provider.

**Comment by rxaviers** [comment]
So far we've discussed three distinct roles for download builder (the online tool). Please, correct me if I'm wrong.
 1. Customizes styles (became #20).
 1. Allows usage without having to run a CSS preprocessor.
 1. Concatenates modules and dependencies into a monolithic bundle.

#### Customizes styles

This topic became #20, so let's have any related discussion there. The summary is that a configuration online tool called ThemeRoller will be provided. It will generate the variables file.

#### Allows usage without having to run a CSS preprocessing.

This is the most concerning one IMO if css-chassis cares about having *wide adoption*.

@dmethvin well put these use cases:

> 1) Will this project use a CSS preprocessor for its own internal development? 2) Will the web devs be required to use a (local or online) CSS preprocessor, or is it optional? It sounds like item 1 is a "definitely" but as for item 2 I don't think devs should have to use a preprocessor before they can take advantage of this framework.

Dave is not the only one who thinks developers shouldn't have to use a preprocessor before they can take advantage of this framework. @kristoferjoseph said the same thing considering his experience on @topcoat:

> We also discovered that developers would balk at any and all preprocessors if it was not the one they thought was best, but funny enough they would use a CDN version of the compiled CSS without even looking at it. Having a way to deliver just the CSS regardless of the preprocessing needed is very important. 

Looking for a "market share" about the preprocessor worlds, I found out that **82% don't use SASS or SCSS** (https://css-tricks.com/poll-results-popularity-of-css-preprocessors/), even being one of the top two most used preprocessors. It reinforces the need for distributing plain CSS.

I think that a simple solution that doesn't require the developer to go into a download builder webpage to download the generated CSS (which I believe is a bad habit for maintenance purposes as discussed above) is to simply include CSS processed files on each release. This way developers can use package managers like **bower** or **npm** and use the **CSS** files. Let me explain...

This problem is very analogous to the module loader on JS development, where source code is written using ES6 modules. But, distributed using different flavors (AMD and CommonJS). Similarly, css-chassis could generate `dist/*.css` files and commit them on detached tags for easier and wider adoption. I would expect `dist/grid.css`, `dist/table.css`, `dist/button.css`, `dist/form.css`, etc. Note a direct map between a `.scss` into a `.css` file is not strictly necessary. For example, `scss/typography.scss` + `scss/utilities.scss` + `scss/mixins.scss` + `scss/helpers.scss` could generate `dist/core.css`.

#### Concatenates modules and dependencies into a monolithic bundle

If one cares about developers using monolithic bundles on their applications, it's clear that Download Builder is not the answer. It's pointless encouraging them to use a custom tool for each of the libraries they depend on and ignore their actual need for creating the bundle for their apps. There are several real tools for managing dependencies and for creating the bundles. Let's encourage them to use that.

For the developers that want a fast and simple solution for prototyping and don't want to setup the bundle themselves, prototyping doesn't need optimization. Having multiple stylesheet link tags in their HTML won't be a problem.

**Comment by rxaviers** [comment]
> here at Thomson Reuters we often use an override approach to keep OSS source completely untouched

@mendeza, I agree that keeping source completely untouched is the key for maintainability. Could you explain what's the override approach you use please?

**Comment by sfrisk** [comment]
A download builder isn't an unreasonable suggestion.  Many CSS Frameworks and UI Toolkits use them, and thanks to pre-processors, it doesn't need to be a headache.

For example, take a look at what [Bootstrap](http://getbootstrap.com/) does: 

Now going to the [download page](http://getbootstrap.com/getting-started/) of bootstrap makes it quite easy to download the CSS as is - minified or un-minified. Downloading the source to do things the LESS and Sass way is offered as a secondary option, for those who would prefer a bit more control and don't want to use the CSS as is or the CDN version.  Pre-processors are not required to use Bootstrap - they are merely there for the more advanced developers who want more control and flexibility.

However, if someone doesn't want the default CSS as it can be found in the CDN, or deal with LESS or Sass, then they have the option of using the [customized download builder](http://getbootstrap.com/customize/).

On this page you have the option of selecting which components you wish to include with your download as well as setting many of the base variable colors.   Thanks to pre-processors like Sass or LESS, it's relatively easy to set variables to set up [if() statements](http://sass-lang.com/documentation/file.SASS_REFERENCE.html#_8) to chose  to include certain components to compile.  Then you just need to run the the pre-processor server side and give the user back the CSS that they require.  This allows them to have some of the power offered by using a pre-processor on their computer, but without having to figure out how to use the pre-processor.  This allows bootstrap to become more accessible for more junior developers and easier to theme.

I'm using Bootstrap as an example here, mostly because of its popularity, and I haven't seen people having a problem with this approach.  The custom builder allows for the variables in the pre-processor to be manipulated in the browser, without having to deal with setting up a pre-processor locally.

Also, I don't think using an article from 2.5 years ago arguing that 82% of market share don't use a pre-processors works in this situation. Front end development has changed a lot in the past 2.5 years and will continue to change. Even if everyone isn't using pre-processor locally, it doesn't mean we shouldn't use one to make it easier to generate the CSS we need  for users.  If anything, it makes it much easier to make Chassis accessible for not only senior but junior front end developers.
